crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +519,293 @@ where
         }
     }
+
+    // Evaluate the result of a batch. Commit or drop the storage changes,
+    // update stats and event, manage replay protection.
+    fn evalute_tx_result(
+        &mut self,
+        response: &mut shim::response::FinalizeBlock,
+        tx_result: Result<namada::tx::data::TxResult<protocol::Error>>,
+        TxData {
+            is_atomic_batch,
+            tx_header,
+            commitments_len,
+            tx_index,
+            replay_protection_hashes,
+            consumed_gas,
+            height,
+            wrapper_args,
+        }: TxData,
+        TxInfo {
+            tx_event,
+            stats,
+            changed_keys,
+        }: TxInfo,
+    ) {
+        match tx_result {
+            Ok(tx_result) => {
+                // Track the need to commit the batch hash for replay
+                // protection. Hash must be written if at least one of the
+                // txs in the batch requires so
+                let mut commit_batch_hash = false;
+                // Atomic batches need custom handling when even a single tx
+                // fails, since we need to drop everything
+                if is_atomic_batch
+                    && tx_result.batch_results.0.iter().any(|(_, result)| {
+                        match result {
+                            Ok(res) => {
+                                // If an inner tx failed for any reason but
+                                // invalid signature, commit its hash to
+                                // storage,
+                                // otherwis allow for a replay
+                                if !res
+                                    .vps_result
+                                    .status_flags
+                                    .contains(VpStatusFlags::INVALID_SIGNATURE)
+                                {
+                                    commit_batch_hash = true;
+                                }
+
+                                !res.is_accepted()
+                            }
+                            Err(e) => {
+                                // If inner transaction didn't fail
+                                // because of invalid
+                                // section commitment, commit its hash to
+                                // prevent replays
+                                if matches!(
+                                    tx_header.tx_type,
+                                    TxType::Wrapper(_)
+                                ) && !matches!(
+                                    e,
+                                    protocol::Error::MissingSection(_)
+                                ) {
+                                    commit_batch_hash = true;
+                                }
+                                true
+                            }
+                        }
+                    })
+                {
+                    // TODO: nee to increment the rejected txs too
```

Comment:
```suggestion
                    // TODO: need to increment the rejected txs too
```

Reply:
btw for this I guess you can do another pass over the batch of tx results. hm but this has raised another question: if the tx is atomic, shouldn't we be short circuiting execution? we don't need to exec all txs